completions/bash/multipass

```diff
@@ -40,10 +40,126 @@ _multipass_complete()
         done
     }
 
-    local cur cmd opts prev prev_opts
+    # Set $opts to the list of available networks.
+    _multipass_networks()
+    {
+        local cmd="multipass networks --format=csv 2>/dev/null"
+
+        opts=$( \eval $cmd | \grep -Ev '(\+--|Name)' | \cut -d',' -f 1 )
+    }
+
+    # Removes the comma or equals sign at the end of $cur.
+    _remove_trailing_char()
+    {
+        if [[ $cur =~ [,=]$ ]]; then
+            cur="${cur#*[,=]}"
+            return 0
+        fi
+
+        return 1
+    }
+
+    # When options for --network are specified in the form var=val, this function sets $opts to the values needed
+    # depending on the value of var (which is specified as the parameter $1).
+    _complete_network_var()
+    {
+        case "$1" in
+            "id")
+                _multipass_networks
+            ;;
+            "mac")
+                opts=""
+            ;;
+            "mode")
+                opts="auto manual"
+            ;;
+        esac
+    }
+
+    # Find the position of the last appearance of a given parameter name in the command-line. Return its index, or
+    # 0 if not found.
+    _find_last_par()
+    {
+        local last_pos
+
+        for ((last_pos=COMP_CWORD; last_pos>0; --last_pos)); do
+            if [[ "${COMP_WORDS[$last_pos]}" == "$1" ]]; then
+                break
+            fi
+        done
+
+        return ${last_pos}
+    }
+
+    # Find the position of the last appearance of parameter starting with "--" in the command-line. Return its
+    # index, or 0 if not found.
+    _find_last_dashdash()
+    {
+        local last_pos
+
+        for ((last_pos=COMP_CWORD; last_pos>0; --last_pos)); do
+            if [[ "${COMP_WORDS[$last_pos]}" =~ ^--.* ]]; then
+                break
+            fi
+        done
+
+        return ${last_pos}
+    }
+
+    # Set $last_dashdash to the name of the last argument whose name starts with "--", or the empty string if there
+    # is none.
+    _last_dashdash_name()
+    {
+        local last_pos
+
+        last_dashdash_arg=""
+        for ((last_pos=COMP_CWORD; last_pos>0; --last_pos)); do
+            if [[ "${COMP_WORDS[$last_pos]}" =~ ^--.* ]]; then
+                last_dashdash_arg=${COMP_WORDS[$last_pos]}
+                break
+            fi
+        done
+    }
```

Comment:
These 3 functions have practically identical code, I think they should be merged. The last two are simple: just add the `return`. The first one could do with a regex match as well, instead of the equality test. So something like (untested):

```bash
    _find_last_match()
    {
        local last_pos

        last_match=""
        for ((last_pos=COMP_CWORD; last_pos>0; --last_pos)); do
            if [[ "${COMP_WORDS[$last_pos]}" =~ "$1" ]]; then
                last_match=${COMP_WORDS[$last_pos]}
                break
            fi
        done

        return ${last_pos}
    }
```

You could then call it only once with "^--.*", check if the resulting match was `== "--network"`, still use "last_pos" to iterate, and also let `_unspecified_network_vars()` use those variables instead of repeating the same thing.

Reply:
Actually, I think make that regex "^-.*", so that short options are also caught. For example, I now get:

```
$ multipass launch --network -n=sadf,
id=    mac=   mode=
```

Not super important, but might as well account for it.

Reply:
Done. You're absolutely right, it was essentially the same function repeated three times. Corrected.

Reply:
The last step is to have only one call instead of three. IIUC you should be able to get all the info you need with a single `_find_last_match "^-.*"` (since you can then check if `last_match == "--networks"`).

Reply:
Done.